### PR TITLE
core: pm: add unregister_pm_cb() + stm32mp257f-ev1 GPIO and console support

### PR DESCRIPTION
**Pull request by GseoC**
This P-R implements:
-  The mean to unregister a PM callback
- GPIO and console support for stm32mp257f-ev1 platform

### CODE REVIEW

File: core/include/kernel/pm.h [hunk 1]
```diff
@@ -169,8 +195,23 @@ static inline void register_pm_core_service_cb(pm_callback callback,
...
+/*
+ * Unregister a core service callback from generic suspend/resume sequences
+ *
+ * @callback: Registered callback function
+ * @handle: Private handle argument that was registered for the callback
+ */
+static inline void unregister_pm_core_service_cb(pm_callback callback,
+						 void *handle)
+{
+	enum pm_callback_order order = PM_CB_ORDER_CORE_SERVICE;
```
Comment (jforissier) [inline core/include/kernel/pm.h:211]: Local variable not needed

**Review by jforissier — COMMENTED** [review]
- For "drivers: stm32_rif: tag unused parameters as __unused":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "core: pm: add unregister_pm_cb()":
  One comment, then:
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/include/kernel/pm.h (hunk 1 above)
Comment (GseoC) [reply core/include/kernel/pm.h:211]: It was to keep alignment with parentheses without exceeding the character per lign limit. Do you suggest I change it nevertheless?

Comment (jforissier) [reply core/include/kernel/pm.h:211]: OK then it's fine as it is. In fact `PM_CALLBACK_HANDLE_INITIALIZER` is too long, it should have been called `PM_CB_HANDLE_INITIALIZER`.

File: core/drivers/stm32_gpio.c [hunk 2]
```diff
@@ -606,6 +650,161 @@ static bool bank_is_registered(const void *fdt, int node)
...
+				goto out;
+			}
+		}
+	}
+
+	/*
+	 * Lock RIF configuration if configured. This cannot be undone until
+	 * next reset.
+	 */
+	io_clrsetbits32(bank->base + GPIO_RCFGLOCKR_OFFSET, GPIO_RCFGLOCKR_MASK,
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:817]: Maybe `io_setbits32()` makes more sense regarding such lock bits.

File: core/drivers/stm32_gpio.c [hunk 3]
```diff
@@ -649,15 +854,11 @@ static TEE_Result dt_stm32_gpio_bank(const void *fdt, int node,
 		panic("missing reg size property");
 
 	DMSG("Bank name %s", fdt_get_name(fdt, node, NULL));
+	bank->base = io_pa_or_va_secure(&pa_va, blen);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:948]: Can remove this line. `bank->base` is set later (lines 879/884/892).

File: core/include/kernel/pm.h [hunk 4]
```diff
@@ -139,6 +139,19 @@ struct pm_callback_handle {
  */
 void register_pm_cb(struct pm_callback_handle *pm_handle);
 
+/*
+ * Register a callback for suspend/resume sequence
```
Comment (etienne-lms) [inline core/include/kernel/pm.h:143]: `Unregister a previously registered PM callback handle`

File: core/include/kernel/pm.h [hunk 5]
```diff
@@ -139,6 +139,19 @@ struct pm_callback_handle {
...
+ * Refer to struct pm_callback_handle for description of the callbacks
+ * API and the registration directives.
+ *
+ * @pm_handle: Reference previously registered
+ *
+ * This function removes from the list of the PM called callbacks the
+ * entry that match callback function reference and private handle or
```
Comment (etienne-lms) [inline core/include/kernel/pm.h:150]: s/or/and/

File: core/drivers/stm32_gpio.c [hunk 6]
```diff
@@ -187,6 +188,18 @@ struct stm32_gpio_bank {
 	STAILQ_ENTRY(stm32_gpio_bank) link;
 };
 
+/*
+ * struct stm32_gpios_pm_state - Consumed GPIO for PM purpose
+ * @gpio_pinctrl: Reference and configuration state for a consumed GPIO
+ * @level: GPIO level
+ * @link: Link consumed GPIO list
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:195]: s/Link consumed/Link in consumed/

File: core/drivers/stm32_gpio.c [hunk 7]
```diff
@@ -321,11 +336,35 @@ static void stm32_gpio_set_direction(struct gpio_chip *chip,
...
-	free(gpio);
+
+	exceptions = cpu_spin_lock_xsave(&gpio_lock);
+
+	SLIST_FOREACH_SAFE(state, &consumed_gpios_head, link, tstate) {
+		if (state->gpio_pinctrl.bank == bank->bank_id &&
+		    state->gpio_pinctrl.pin == gpio->pin) {
+			SLIST_REMOVE(&consumed_gpios_head, state,
+				     stm32_gpio_pm_state, link);
+			unregister_pm_driver_cb(consumed_gpios_pm, state);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:359]: `register_pm_core_service_cb()` to match line 1342.

File: core/drivers/stm32_gpio.c [hunk 8]
```diff
@@ -767,6 +842,22 @@ static TEE_Result apply_rif_config(struct stm32_gpio_bank *bank)
...
+static void stm32_gpio_save_rif_config(struct stm32_gpio_bank *bank)
+{
+	size_t i = 0;
+
+	for (i = 0; i < bank->ngpios; i++)
+		bank->rif_cfg->cid_confs[i] = io_read32(bank->base +
+							 GPIO_CIDCFGR(i));
+
+	bank->rif_cfg->priv_conf[0] = io_read32(bank->base +
+						 GPIO_PRIVCFGR_OFFSET);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:854]: indentaion

File: core/include/kernel/pm.h (hunk 1 above)
Comment (etienne-lms) [reply core/include/kernel/pm.h:211]: By the way, we could also shorten `struct pm_callback_handle` to `struct pm_cb_handle` :)

File: core/drivers/stm32_gpio.c [hunk 9]
```diff
@@ -606,6 +650,161 @@ static bool bank_is_registered(const void *fdt, int node)
...
+		/* Check if the controller is in semaphore mode */
+		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1))
+			continue;
+
+		/* If not TDCID, we want to acquire semaphores assigned to us */
+		res = stm32_rif_acquire_semaphore(bank->base + GPIO_SEMCR(i),
+						  GPIO_MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Could not acquire semaphore for pin %u", i);
+			goto out;
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_gpio.c:764]: It should return with an error code, I corrected this.

File: core/drivers/stm32_gpio.c (hunk 2 above)
Comment (GseoC) [reply core/drivers/stm32_gpio.c:817]: Yes, I agree

File: core/drivers/stm32_gpio.c (hunk 7 above)
Comment (GseoC) [reply core/drivers/stm32_gpio.c:359]: changed all to driver callbacks

**Comment by GseoC** [comment]
Updated with comments applied, I will add all tags at the end of the review

File: core/drivers/stm32_gpio.c (hunk 7 above)
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:360]: +1

File: core/drivers/stm32_gpio.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:764]: +1

File: core/drivers/stm32_gpio.c [hunk 10]
```diff
@@ -933,21 +1209,146 @@ static TEE_Result stm32_pinctrl_dt_get(struct dt_pargs *pargs,
...
+
+	STAILQ_FOREACH(bank, &bank_list, link) {
+		if (bank->rif_cfg) {
+			if (!bank->is_tdcid)
+				continue;
+
+			bank->rif_cfg->access_mask[0] = GENMASK_32(bank->ngpios,
+								   0);
+
+			apply_rif_config(bank);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:1242]: I think we should test the return value here, to address case when for example we fail to get a semaphore we need.

File: core/drivers/stm32_gpio.c [hunk 11]
```diff
@@ -933,21 +1209,146 @@ static TEE_Result stm32_pinctrl_dt_get(struct dt_pargs *pargs,
...
+			continue;
+
+		if (bank->rif_cfg) {
+			res = apply_rif_config(bank);
+			if (res) {
+				EMSG("Failed to set GPIO RIF, bank %u",
+				     bank->bank_id);
+				STAILQ_REMOVE(&bank_list, bank, stm32_gpio_bank,
+					      link);
+				free(bank);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:1308]: `bank` holds various pointers that would also need to be free. I think we could simply return with`res` here at let caller deal with it.

**Comment by GseoC** [comment]
Comments addressed

File: core/drivers/stm32_gpio.c [hunk 12]
```diff
@@ -141,26 +181,46 @@ struct stm32_gpio_bank {
...
+	struct rif_conf_data *rif_cfg;
+	uint32_t seccfgr;
 	bool sec_support;
+	bool ready;
+	bool is_tdcid;
 	STAILQ_ENTRY(stm32_gpio_bank) link;
 };
 
+/*
+ * struct stm32_gpios_pm_state - Consumed GPIO for PM purpose
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:193]: typo: s/stm32_gpios_pm_state/stm32_gpio_pm_state/

File: core/drivers/stm32_gpio.c [hunk 13]
```diff
@@ -141,26 +181,46 @@ struct stm32_gpio_bank {
...
+	uint8_t level;
+	SLIST_ENTRY(stm32_gpio_pm_state) link;
+};
+
 /**
  * Compatibility information of supported banks
  *
  * @gpioz: True if bank is a GPIOZ bank
  * @secure_control: Identify GPIO security bank capability.
+ * @secure_control: Identify RIF presence.
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:209]: typo: s/secure_control/secure_extended/

File: core/drivers/stm32_gpio.c [hunk 14]
```diff
@@ -606,6 +725,178 @@ static bool bank_is_registered(const void *fdt, int node)
...
+
+		/* Check if the controller is in semaphore mode */
+		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1))
+			continue;
+
+		/* If not TDCID, we want to acquire semaphores assigned to us */
+		res = stm32_rif_acquire_semaphore(bank->base + GPIO_SEMCR(i),
+						  GPIO_MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Could not acquire semaphore for pin %u", i);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:763]: Suggestion: print also the GPIO bank ID to ease debugging. Eg.
E.g. replace `for pin %u", i)`
with `for pin %C%u", 'A' + bank->bank_id, i)`

This comment applies also to lines 797 and 806.

File: core/drivers/stm32_gpio.c [hunk 15]
```diff
@@ -606,6 +725,178 @@ static bool bank_is_registered(const void *fdt, int node)
...
+	bank->rif_cfg = calloc(1, sizeof(*bank->rif_cfg));
+	if (!bank->rif_cfg)
+		panic();
+
+	bank->rif_cfg->sec_conf = calloc(1, sizeof(uint32_t));
+	if (!bank->rif_cfg->sec_conf)
+		panic();
+
+	nb_rif_conf = MIN((unsigned int)(lenp / sizeof(uint32_t)),
+			  bank->ngpios);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:885]: If think we should rather report an error if `lenp / sizeof(uint32_t) > bank->ngpios`.

File: core/drivers/stm32_gpio.c [hunk 16]
```diff
@@ -634,6 +925,12 @@ static TEE_Result dt_stm32_gpio_bank(const void *fdt, int node,
 	if (!bank)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
+	if (compat->secure_extended) {
+		res = stm32_rifsc_check_tdcid(&bank->is_tdcid);
+		if (res)
+			return res;
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:931]: `free(bank)`

**Review by etienne-lms — COMMENTED** [review]
Some late comments.
I'm fine you fix and squash all fixup commits. I'll make a last review run afterward.

File: core/drivers/stm32_gpio.c (hunk 14 above)
Comment (GseoC) [reply core/drivers/stm32_gpio.c:763]: Done with few other occurences as well

File: core/drivers/stm32_gpio.c (hunk 15 above)
Comment (GseoC) [reply core/drivers/stm32_gpio.c]: I put an assert but let me know if an error case handling feels better to you

**Comment by GseoC** [comment]
Comments addressed

**Comment by GseoC** [comment]
@etienne-lms, any more comments? :)

File: core/drivers/stm32_gpio.c [hunk 17]
```diff
@@ -606,6 +651,163 @@ static bool bank_is_registered(const void *fdt, int node)
...
+	}
+
+	/* Security and privilege RIF configuration */
+	io_clrsetbits32(bank->base + GPIO_PRIVCFGR_OFFSET, GPIO_PRIVCFGR_MASK,
+			bank->rif_cfg->priv_conf[0]);
+	io_clrsetbits32(bank->base + GPIO_SECR_OFFSET, GPIO_SECCFGR_MASK,
+			bank->rif_cfg->sec_conf[0]);
+
+	if (!bank->is_tdcid)
+		goto out;
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:776]: For consistency, I think we should add `res = TEE_SUCCESS` here.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: stm32_rif: tag unused parameters as __unused",
"core: pm: add unregister_pm_cb()",
"drivers: stm32_gpio: save/restore consumed GPIOs in PM sequence",
"dts: stm32: add console support on USART2 for stm32mp257f-ev1" and
"dts: stm32_gpio: add GPIO banks RIF configurations for stm32mp257f-ev1".

Some minor comments for commit "drivers: stm32_gpio: add stm32mp25x support".

File: core/drivers/stm32_gpio.c [hunk 18]
```diff
@@ -12,7 +12,7 @@
 #include <drivers/gpio.h>
 #include <drivers/pinctrl.h>
 #include <drivers/stm32_gpio.h>
-#include <dt-bindings/gpio/stm32mp_gpio.h>
+#include <drivers/stm32_rif.h>
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_gpio.c:15]: Rather: #include <dt-bindings/gpio/gpio.h>

**Comment by GseoC** [comment]
Comments addressed and tags applied, commit `drivers: stm32_gpio: add stm32mp25x support` still needs one last review

File: core/drivers/stm32_gpio.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:15]: Prefer _dt-bindings/gpio/stm32mp_gpio.h_. It defines some STM32 specific binding information needed (`TZPROT()` macro) and includes dt-_bindings/gpio/gpio.h_.

[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_gpio.c:15]: TZPROT use is removed by this patchset

[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:15]: Indeed, sorry. Fine with _dt-bindings/gpio/gpio.h_.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"drivers: stm32_gpio: add stm32mp25x support".

**Comment by GseoC** [comment]
Tags applied and last patch squashed, thanks

**Comment by GseoC** [comment]
Pushed again because of failed PKCS11 1007 failed test

**Review by etienne-lms — APPROVED** [review]

**Pull request merged**